book-content/chapters/02-ide-superpowers.md

```diff
@@ -0,0 +1,579 @@
+dragleave
+...
+```
+
+Autocomplete is an essential tool for writing TypeScript code, and it's available right out of the box in VS Code.
+
+### Exercises
+
+#### Exercise 1: Autocomplete
+
+Here's an example of some code where autocomplete can be triggered:
+
+```typescript
+const acceptsObj = (obj: { foo: string; bar: number; baz: boolean }) => {};
+
+acceptsObj({
+  // Autocomplete in here!
+});
+```
+
+Practice using the autocomplete shortcut to fill in the object when calling `acceptsObj`.
+
+<Exercise title="Exercise 1: Autocomplete" filePath="/src/016.5-ide-superpowers/044-manually-triggering-autocomplete.problem.ts"></Exercise>
+
+#### Solution 1: Autocomplete
+
+When you hit `Ctrl + Space` inside the object, you'll see a list of the possible properties based on the `MyObj` type:
+
+```typescript
+bar;
+baz;
+foo;
+```
+
+As you select each property, the autocomplete list will update to show you the remaining properties.
+
+## TypeScript Error Checking
+
+The thing TypeScript is most famous for is its errors. These are a set of rules which TypeScript uses to make sure your code is doing what you think it's doing.
+
+For every change you make to a file, the TypeScript server will check your code.
+
+If the TypeScript server finds any errors, it will tell VS Code to draw a red squiggly line under the part of the code that has a problem. Hovering over the underlined code will show you the error message. Once you make a change, the TypeScript server will check again and remove the red squiggly line if the error is fixed.
+
+I like to think of it like a teacher hovering over your shoulder, underlining your work in red pen while you type.
+
+Let's look at these errors a bit more deeply.
+
+### Catching Runtime Errors
+
+Sometimes TypeScript will warn you about things that will definitely fail at runtime.
+
+For example, consider the following code:
+
+```ts twoslash
+// @errors: 18047
+const a = null;
+
+a.toString();
+```
+
+TypeScript tells us that there is a problem with `a`. This tells us where the problem is, but it doesn't necessarily tell us what the problem is. In this case, we need to stop and think about why we can't call `toString()` on `null`. If we do, it will throw an error at runtime.
+
+```
+Uncaught TypeError: Cannot read properties of null (reading 'toString').
+```
+
+Here, TypeScript is telling us that an error might happen even without us needing to check it. Very handy.
+
+### Warnings About Non-Runtime Errors
+
+Not everything TypeScript warns us about will actually fail at runtime.
+
+Take a look at this example where we're assigning a property to an empty object:
+
+```ts twoslash
+// @errors: 2339
+const obj = {};
+
+const result = obj.foo;
+```
+
+TypeScript draws a red squiggly line below `foo`. But if we think about it, this code won't actually cause an error at runtime. We're trying to assign a property that doesn't exist in this object: `foo`. This won't error, it will just mean that result is undefined.
```

Comment:
```suggestion
TypeScript draws a red squiggly line below `foo`. But if we think about it, this code won't actually cause an error at runtime. We're trying to read a property that doesn't exist in this object: `foo`. This won't error, it will just mean that result is undefined.
```